### PR TITLE
Checks URLs against an URL array using the Cake router

### PR DESCRIPTION
**Pull request by burzum**
Refs https://github.com/cakephp/authentication/issues/167

### CODE REVIEW

File: tests/TestCase/UrlChecker/CakeRouterCheckerTest.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     * @return void
+     */
+    public function testCheckFullUrls()
+    {
+        $checker = new CakeRouterChecker();
+        $request = ServerRequestFactory::fromGlobals(
+            ['REQUEST_URI' => '/users/login']
+        );
+
+        $result = $checker->check($request, '/users/login', [
```
Comment (robertpustulka) [inline tests/TestCase/UrlChecker/CakeRouterCheckerTest.php]: Shouldn't there be a Router url array?

Comment (burzum) [reply tests/TestCase/UrlChecker/CakeRouterCheckerTest.php]: Fixed.

**Comment by burzum** [comment]
@robertpustulka what do you think about this? https://gist.github.com/burzum/5decfd9147a243d55e117d224ce313ae

Instead of creating a complete new checker and duplicating two methods from the default checker, why not just extending it the way I'm doing it?

**Comment by lilHermit** [comment]
I don't see the point of the `checkFullUrl` option as both urls are derived from `Router::url` it's not like you're  comparing a string from the developer? Unless I'm missing something!

File: src/UrlChecker/CakeRouterChecker.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function check(ServerRequestInterface $request, $loginUrls, array $options = [])
+    {
+        $options = $this->_mergeDefaultOptions($options);
+        $url = $this->_getUrlFromRequest($request->getUri(), $options['checkFullUrl']);
+
+        // Support string loginUrls
+        if (is_string($loginUrls)) {
```
Comment (markstory) [inline src/UrlChecker/CakeRouterChecker.php]: Why? I thought the goal of this checker was to use the router features?

File: tests/TestCase/UrlChecker/CakeRouterCheckerTest.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+        $checker = new CakeRouterChecker();
+        $request = ServerRequestFactory::fromGlobals(
+            ['REQUEST_URI' => '/users/invalid']
+        );
+        $result = $checker->check($request, [
+            'controller' => 'Users',
+            'action' => 'login'
+        ]);
+        $this->assertFalse($result);
```
Comment (markstory) [inline tests/TestCase/UrlChecker/CakeRouterCheckerTest.php]: Should there be a test for a named route?

File: src/UrlChecker/CakeRouterChecker.php (hunk 2 above)
Comment (burzum) [reply src/UrlChecker/CakeRouterChecker.php]: @markstory thats a question for @lilHermit

File: tests/TestCase/UrlChecker/CakeRouterCheckerTest.php (hunk 3 above)
Comment (burzum) [reply tests/TestCase/UrlChecker/CakeRouterCheckerTest.php]: OK

File: src/UrlChecker/CakeRouterChecker.php (hunk 2 above)
Comment (robertpustulka) [reply src/UrlChecker/CakeRouterChecker.php]: I also don't think that is necessary. If someone needs to support both they can implement a checker collection.

File: tests/TestCase/UrlChecker/CakeRouterCheckerTest.php [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            ['REQUEST_URI' => '/users/login']
+        );
+        $result = $checker->check($request, $url, [
+            'checkFullUrl' => true
+        ]);
+        $this->assertTrue($result);
+
+        $checker = new CakeRouterChecker();
+        $request = ServerRequestFactory::fromGlobals(
+            ['REQUEST_URI' => '/users/invalid']
```
Comment (robertpustulka) [inline tests/TestCase/UrlChecker/CakeRouterCheckerTest.php]: I think we need a test for url with different hosts to make sure this feature works.

File: docs/URL-Checkers.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# URL Checkers
```
Comment (robertpustulka) [inline docs/URL-Checkers.md]: Checkers are already mentioned here: https://github.com/cakephp/authentication/blob/master/docs/Authenticators.md#url-checkers

You may want to link this from there or remove checker docs from Authenticators page.

File: src/UrlChecker/CakeRouterChecker.php (hunk 2 above)
Comment (lilHermit) [reply src/UrlChecker/CakeRouterChecker.php]: I just added it because thought if it was a string its easy to compare but happy for it to fail

**Comment by burzum** [comment]
@markstory @robertpustulka can we merge it now?

File: src/UrlChecker/CakeRouterChecker.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+        // If it's a single route array add to another
+        if (!is_numeric(key($loginUrls))) {
+            $loginUrls = [$loginUrls];
+        }
+
+        foreach ($loginUrls as $validUrl) {
+            try {
+                $validUrl = Router::url($validUrl, $options['checkFullUrl']);
+            } catch (MissingRouteException $e) {
```
Comment (robertpustulka) [inline src/UrlChecker/CakeRouterChecker.php]: I don’t think we should suppress this exception and allow for missing routes silently.

File: src/UrlChecker/CakeRouterChecker.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function check(ServerRequestInterface $request, $loginUrls, array $options = [])
+    {
+        $options = $this->_mergeDefaultOptions($options);
+        $url = $this->_getUrlFromRequest($request->getUri(), $options['checkFullUrl']);
+
+        if (!is_array($loginUrls) || empty($loginUrls)) {
```
Comment (robertpustulka) [inline src/UrlChecker/CakeRouterChecker.php]: Default checker returns true for empty url. Should we allow for the same behavior here?

Comment (burzum) [reply src/UrlChecker/CakeRouterChecker.php]: Why does it return true? Shouldn't we always return false if it doesn't match or is empty?

Comment (robertpustulka) [reply src/UrlChecker/CakeRouterChecker.php]: I’m not sure why. It had been that way in the FormAuthentication, before checkers were implemented.

Comment (burzum) [reply src/UrlChecker/CakeRouterChecker.php]: I'm OK with either way we just need to make a decision. :) But wouldn't it make more sense to return false if it's empty?

Comment (robertpustulka) [reply src/UrlChecker/CakeRouterChecker.php]: It would probably be safer that way. So I’m ok with returning false on empty urls.

Comment (burzum) [reply src/UrlChecker/CakeRouterChecker.php]: Are we OK to merge this then?

Comment (robertpustulka) [reply src/UrlChecker/CakeRouterChecker.php]: I think we need another test as I mentioned here: https://github.com/cakephp/authentication/pull/171/files#r157144178

**Pull request merged**